sugar.go

```diff
@@ -0,0 +1,214 @@
+	if ce := s.core.Check(ErrorLevel, sweetenMsg(msg)); ce != nil {
+		ce.Write(sweetenFields(keysAndValues, s.core)...)
+	}
+}
+
+// Errorf uses fmt.Sprintf to construct a dynamic message and logs it at
+// ErrorLevel. It doesn't add to the message's structured context.
+func (s *SugaredLogger) Errorf(template string, args ...interface{}) {
+	if ce := s.core.Check(ErrorLevel, fmt.Sprintf(template, args...)); ce != nil {
+		ce.Write()
+	}
+}
+
+// DPanic logs a message and some key-value pairs using the underlying logger's
+// DPanic method. Keys and values are treated as they are in the With
+// method. (See Logger.DPanic for details.)
+func (s *SugaredLogger) DPanic(msg interface{}, keysAndValues ...interface{}) {
+	if ce := s.core.Check(DPanicLevel, sweetenMsg(msg)); ce != nil {
+		ce.Write(sweetenFields(keysAndValues, s.core)...)
+	}
+}
+
+// DPanicf uses fmt.Sprintf to construct a dynamic message, which is passed to
+// the underlying Logger's DPanic method. (See Logger.DPanic for details.) It
+// doesn't add to the message's structured context.
+func (s *SugaredLogger) DPanicf(template string, args ...interface{}) {
+	if ce := s.core.Check(DPanicLevel, fmt.Sprintf(template, args...)); ce != nil {
+		ce.Write()
+	}
+}
+
+// Panic logs a message and some key-value pairs at PanicLevel, then panics.
+// Keys and values are treated as they are in the With method.
+func (s *SugaredLogger) Panic(msg interface{}, keysAndValues ...interface{}) {
+	if ce := s.core.Check(PanicLevel, sweetenMsg(msg)); ce != nil {
+		ce.Write(sweetenFields(keysAndValues, s.core)...)
+	}
+}
+
+// Panicf uses fmt.Sprintf to construct a dynamic message and logs it at
+// PanicLevel, then panics. It doesn't add to the message's structured context.
+func (s *SugaredLogger) Panicf(template string, args ...interface{}) {
+	if ce := s.core.Check(PanicLevel, fmt.Sprintf(template, args...)); ce != nil {
+		ce.Write()
+	}
+}
+
+// Fatal logs a message and some key-value pairs at FatalLevel, then calls
+// os.Exit(1). Keys and values are treated as they are in the With method.
+func (s *SugaredLogger) Fatal(msg interface{}, keysAndValues ...interface{}) {
+	if ce := s.core.Check(FatalLevel, sweetenMsg(msg)); ce != nil {
+		ce.Write(sweetenFields(keysAndValues, s.core)...)
+	}
+}
+
+// Fatalf uses fmt.Sprintf to construct a dynamic message and logs it at
+// FatalLevel, then calls os.Exit(1). It doesn't add to the message's
+// structured context.
+func (s *SugaredLogger) Fatalf(template string, args ...interface{}) {
+	if ce := s.core.Check(FatalLevel, fmt.Sprintf(template, args...)); ce != nil {
+		ce.Write()
+	}
+}
+
+func sweetenFields(args []interface{}, errLogger Logger) []zapcore.Field {
+	if len(args) == 0 {
+		return nil
+	}
+	if len(args)%2 == 1 {
+		errLogger.DPanic(oddNumberErrMsg, Any("ignored", args[len(args)-1]))
+	}
+
+	fields := make([]zapcore.Field, len(args)/2)
+	for i := range fields {
+		key := sweetenMsg(args[2*i])
```

Comment:
I've had a number of very long discussions about this, and I'm generally convinced that there's utility here. We already have a logger that's type-safe, low-allocation, and fast. For the sugared logger, I think there's value in supporting `error` and `fmt.Stringer` in addition to strings. Given the limitations of the type system and reasonable APIs, eating an allocation and accepting `interface{}` seems like a reasonable tradeoff.